server/schedule/placement/config.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package placement
+
+import (
+	"bytes"
+	"encoding/json"
+)
+
+// ruleConfig contains rule and rule group configurations.
+type ruleConfig struct {
+	rules  map[[2]string]*Rule   // {group, id} => Rule
+	groups map[string]*RuleGroup // id => RuleGroup
+}
+
+func newRuleConfig() *ruleConfig {
+	return &ruleConfig{
+		rules:  make(map[[2]string]*Rule),
+		groups: make(map[string]*RuleGroup),
+	}
+}
+
+// adjust configs for `buildRuleList` and API use.
+func (c *ruleConfig) adjust() {
+	// remove all default group configurations.
+	// if there are rules belong to the group, it will be re-add later.
+	for id, g := range c.groups {
+		if g.isDefault() {
+			delete(c.groups, id)
+		}
+	}
+	for _, r := range c.rules {
+		g := c.groups[r.GroupID]
+		if g == nil {
+			// create default group configurations.
+			g = &RuleGroup{ID: r.GroupID}
+			c.groups[r.GroupID] = g
+		}
+		// setup group for `buildRuleList`
+		r.group = g
+	}
+}
+
+func (c *ruleConfig) getRule(key [2]string) *Rule {
+	return c.rules[key]
+}
+
+func (c *ruleConfig) iterateRules(f func(*Rule)) {
+	for _, r := range c.rules {
+		f(r)
+	}
+}
+
+func (c *ruleConfig) setRule(r *Rule) {
+	c.rules[r.Key()] = r
+}
+
+func (c *ruleConfig) beginPatch() *ruleConfigPatch {
+	return &ruleConfigPatch{
+		c:   c,
+		mut: newRuleConfig(),
+	}
+}
+
+// A helper data structure to update ruleConfig.
+type ruleConfigPatch struct {
+	c   *ruleConfig
```

Comment:
Does `c` mean the existed rules before committing the patching? Please add a comment to explain what `c` works for?

Reply:
ok